pkg/cluster/manager/deploy.go

```diff
@@ -275,15 +271,16 @@ func (m *Manager) Deploy(
 			// copy dependency component if needed
 			switch inst.ComponentName() {
 			case spec.ComponentTiSpark:
-				env := environment.GlobalEnv()
-				var sparkVer utils.Version
-				if sparkVer, _, iterErr = env.V1Repository().WithOptions(repository.Options{
-					GOOS:   inst.OS(),
-					GOARCH: inst.Arch(),
-				}).LatestStableVersion(spec.ComponentSpark, false); iterErr != nil {
-					return
+				// Get spark version and TiSpark version from master or worker
+				var sparkVersion, tiSparkVersion, scalaVersion string
+				if len(topo.(*spec.Specification).TiSparkMasters) > 0 {
+					sparkVersion, tiSparkVersion, scalaVersion = topo.(*spec.Specification).TiSparkMasters[0].GetVersion()
+				} else {
```

Comment:
Is this possible? I think we have a check in spec that does not allow master-less topology

Reply:
yes, it is impossible in theory. but do you think it is necessary to do this judge for unexpected situations?